protocols/imap/src/main/java/org/apache/james/imap/decode/parser/GetAnnotationCommandParser.java

```diff
@@ -0,0 +1,147 @@
+import org.apache.james.imap.api.ImapConstants;
+import org.apache.james.imap.api.ImapMessage;
+import org.apache.james.imap.api.display.HumanReadableText;
+import org.apache.james.imap.api.process.ImapSession;
+import org.apache.james.imap.decode.ImapRequestLineReader;
+import org.apache.james.imap.decode.ImapRequestLineReader.CharacterValidator;
+import org.apache.james.imap.decode.base.AbstractImapCommandParser;
+import org.apache.james.imap.message.request.GetAnnotationRequest;
+import org.apache.james.imap.message.request.GetAnnotationRequest.Depth;
+import org.apache.james.protocols.imap.DecodingException;
+
+import java.util.Set;
+import java.util.regex.Pattern;
+
+public class GetAnnotationCommandParser extends AbstractImapCommandParser {
+    private static final CharMatcher ENDOFLINE_PATTERN = CharMatcher.isNot('\n').and(CharMatcher.isNot('\r'));
+    private final static byte[] MAXSIZE = "MAXSIZE".getBytes();
+    private final static byte[] DEPTH = "DEPTH".getBytes();
+
+    public GetAnnotationCommandParser() {
+        super(ImapCommand.authenticatedStateCommand(ImapConstants.GETANNOTATION_COMMAND_NAME));
+    }
+
+    @Override
+    protected ImapMessage decode(ImapCommand command, ImapRequestLineReader request, String tag, ImapSession session)
+            throws DecodingException {
+        try {
+            GetAnnotationRequest annotationRequest = new GetAnnotationRequest(tag, command, request.mailbox());
+
+            while (nextNonSpaceChar(request) == '(') {
+                consumeParenthesis(request, annotationRequest);
+            }
+
+            if (ENDOFLINE_PATTERN.matches(nextNonSpaceChar(request))) {
+                consumeKey(request, annotationRequest);
+            }
+
+            Preconditions.checkArgument(annotationRequest.isValidParams());
+            return annotationRequest;
+        } catch (NullPointerException e) {
+            throw new DecodingException(HumanReadableText.ILLEGAL_ARGUMENTS, e.getMessage());
+        } catch (IllegalArgumentException e) {
+            throw new DecodingException(HumanReadableText.ILLEGAL_ARGUMENTS, e.getMessage());
+        }
+    }
+
+    private void consumeParenthesis(ImapRequestLineReader request, GetAnnotationRequest getAnnotationRequest) throws DecodingException {
+        request.consumeChar('(');
+        switch (request.nextChar()) {
+            case 'M':
+                getAnnotationRequest.setMaxsizeValue(consumeMaxsizeOpt(request));
+                break;
+
+            case 'D':
+                getAnnotationRequest.setDepthValue(consumeDepthOpt(request));
+                break;
+
+            default:
+                getAnnotationRequest.setKeys(consumeKeys(request));
+                break;
+        }
+        request.consumeChar(')');
+        if (!getAnnotationRequest.getKeys().isEmpty()) {
+            request.eol();
+        }
+    }
+
+    private Depth consumeDepthOpt(ImapRequestLineReader request) throws DecodingException {
```

Comment:
Wrong naming for ImapRequestLineReader request.

It is not a request. It is confusing.